### PR TITLE
pools: Apply blacklist based on AzureIngressProhibitedTarget CRDs

### PR DESCRIPTION
**Pull request by draychev**
This PR introduces (a partial implementation of) brownfield deployment.
The strategy is described in [a proposal here](https://github.com/Azure/application-gateway-kubernetes-ingress/blob/master/proposals/config_ownership_crd.md).

### Overview 
The overall goal is to define certain targets (think `domain/url`) via the `AzureIngressProhibitedTarget` CRD that are blacklisted, which AGIC will not mutate or overwrite. These will be retained as they exist in App Gateway.

This would allow an administrator to *manually* configure some settings on Application Gateway, while allowing the ingress controller to set others.

### Changes
  - this PR implements target blacklist for backend pools only
  - using feature flag to guard these changes (until they are fully fleshed out and tested): `cbCtx.EnvVariables.EnableBrownfieldDeployment`
  - adding to package `brownfield`:
    - `GetExistingBlacklistedPools` - given a list of existing backend pools (from App Gateway) - get the list of those AGIC should retain as part of the config and should not change
    - `MergePools` - merges list of blacklisted and newly created pools into a new one - to be applied to App Gateway

### Overall strategy:

1. Get the list of existing backend pools from App Gateway
2. From the existing list - get the ones blacklisted - these we will not change - will retain as-is
3. Filter Ingress resource to HTTP rules AGIC is allowed to manage
4. Construct backend pools for managed targets only
5. Merge the output 2 and 4 - apply onto App Gateway

### CODE REVIEW

File: pkg/appgw/configbuilder.go [hunk 1]
```diff
@@ -179,3 +179,32 @@ func (c *appGwConfigBuilder) addTags() {
...
+	if c.appGw.BackendAddressPools == nil {
+		return []n.ApplicationGatewayRequestRoutingRule{}
+	}
+	return *c.appGw.RequestRoutingRules
+}
+
+func (c *appGwConfigBuilder) getExistingPathMaps() []n.ApplicationGatewayURLPathMap {
+	if c.appGw.BackendAddressPools == nil {
+		return []n.ApplicationGatewayURLPathMap{}
+
```
Comment (akshaysngupta) [inline pkg/appgw/configbuilder.go]: spacing

Comment (draychev) [reply pkg/appgw/configbuilder.go]: :)

File: pkg/appgw/configbuilder.go [hunk 2]
```diff
@@ -179,3 +179,31 @@ func (c *appGwConfigBuilder) addTags() {
...
+func (c *appGwConfigBuilder) getExistingListeners() []n.ApplicationGatewayHTTPListener {
+	if c.appGw.HTTPListeners == nil {
+		return []n.ApplicationGatewayHTTPListener{}
+	}
+	return *c.appGw.HTTPListeners
+}
+
+func (c *appGwConfigBuilder) getExistingRoutingRules() []n.ApplicationGatewayRequestRoutingRule {
+	if c.appGw.BackendAddressPools == nil {
```
Comment (akshaysngupta) [inline pkg/appgw/configbuilder.go]: s/BackendAddressPools/ApplicationGatewayRequestRoutingRule

File: pkg/appgw/configbuilder.go [hunk 3]
```diff
@@ -179,3 +179,31 @@ func (c *appGwConfigBuilder) addTags() {
...
+		return []n.ApplicationGatewayRequestRoutingRule{}
+	}
+	return *c.appGw.RequestRoutingRules
+}
+
+func (c *appGwConfigBuilder) getExistingPathMaps() []n.ApplicationGatewayURLPathMap {
+	if c.appGw.BackendAddressPools == nil {
```
Comment (akshaysngupta) [inline pkg/appgw/configbuilder.go]: s/BackendAddressPools/ApplicationGatewayURLPathMap

File: pkg/appgw/configbuilder.go (hunk 2 above)
Comment (draychev) [reply pkg/appgw/configbuilder.go]: Thanks for catching this!

File: pkg/appgw/configbuilder.go (hunk 3 above)
Comment (draychev) [reply pkg/appgw/configbuilder.go]: 🐻 thanks!

File: pkg/brownfield/pools_test.go [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+import (
+	n "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2018-12-01/network"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/tests"
+	"github.com/Azure/application-gateway-kubernetes-ingress/pkg/tests/fixtures"
+)
+
+var _ = Describe("test TargetBlacklist/TargetWhitelist backend pools", func() {
```
Comment (akshaysngupta) [inline pkg/brownfield/pools_test.go]: remove TargetWhitelist

File: pkg/brownfield/pools.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+		if rule.URLPathMap == nil {
+			// SSL Redirects do not have BackendAddressPool
+			if rule.BackendAddressPool == nil {
+				continue
+			}
+			poolName := backendPoolName(utils.GetLastChunkOfSlashed(*rule.BackendAddressPool.ID))
+			poolToTarget[poolName] = append(poolToTarget[poolName], target)
+		} else {
+			// Follow the path map
+			pathMapName := pathmapName(utils.GetLastChunkOfSlashed(*rule.URLPathMap.ID))
```
Comment (akshaysngupta) [inline pkg/brownfield/pools.go]: I think we are missing `pathNameToPath[pathMapName].DefaultBackendPool`

Comment (draychev) [reply pkg/brownfield/pools.go]: Agreed - `DefaultBackendAddressPool` needs to be accounted for here

**Pull request merged**